// vue-writer/dist/vue-writer.js
import { defineComponent as V, ref as u, openBlock as x, createElementBlock as k, renderSlot as w, createElementVNode as d, toDisplayString as B, normalizeClass as C } from "vue";
const E = { class: "is-typed" }, I = { class: "typed" }, W = /* @__PURE__ */ V({
  __name: "vue-writer",
  props: {
    array: {},
    eraseSpeed: { default: 100 },
    typeSpeed: { default: 200 },
    delay: { default: 2e3 },
    intervals: { default: 500 },
    start: { default: 0 },
    caret: { default: "cursor" },
    iterations: { default: 0 }
  },
  emits: ["typed"],
  setup(n, { emit: f }) {
    const {
      array: a,
      eraseSpeed: p,
      typeSpeed: i,
      delay: m,
      intervals: y,
      start: _,
      caret: h,
      iterations: S
    } = n, g = f, r = u(""), c = u(0), e = u(!1), t = u(0), l = u(0);
    function o() {
      let s = 0;
      if (l.value < a[t.value].length)
        e.value || (e.value = !0), r.value += a[t.value].charAt(l.value), l.value += 1, setTimeout(o, i);
      else {
        if (c.value += 1, T(a[t.value]), c.value === a.length && (s += 1, s === S))
          return e.value = !1;
        e.value = !1, setTimeout(v, m);
      }
    }
    function v() {
      l.value > 0 ? (e.value || (e.value = !0), r.value = a[t.value].substring(0, l.value - 1), l.value -= 1, setTimeout(v, p)) : (e.value = !1, t.value += 1, t.value >= a.length && (t.value = 0), setTimeout(o, i + y));
    }
    function T(s) {
      g("typed", s);
    }
    return setTimeout(o, _), (s, b) => (x(), k("div", E, [
      w(s.$slots, "default"),
      d("span", I, B(r.value), 1),
      d("span", {
        class: C(h + " " + { typing: e.value })
      }, " ", 2)
    ]));
  }
}), A = {
  install: (n) => {
    n.component("VueWriter", W);
  }
};
export {
  W as VueWriter,
  A as default
};
